vortex-ffi/src/lib.rs

```diff
@@ -41,3 +46,134 @@ pub(crate) unsafe fn to_string_vec(ptr: *const *const c_char, len: c_int) -> Vec
         .map(|i| unsafe { to_string(*ptr.offset(i as isize)) })
         .collect()
 }
+
+/// Define a native FFI type that wraps an [`std::sync::Arc<T>`] type with unsized T.
+///
+/// To solve the problem of dynamic traits using fat pointers, we box the `Arc<T>` and pass a
+/// pointer to the heap-allocated arc struct.
+///
+/// Note: Box::into_raw produces mut pointers, since in theory you can mutate the contents of the
+/// box. In practice though, our box contains an arc, and the callers of this macro tend to deal
+/// in arcs, and so const pointers make more sense.
+#[macro_export]
+macro_rules! arc_dyn_wrapper {
+    ($(#[$meta:meta])* $T:ty, $ffi_ident:ident) => {
+        paste::paste! {
+            $(#[$meta])*
+            #[allow(non_camel_case_types)]
+            pub struct $ffi_ident(std::sync::Arc<$T>);
+
+            #[allow(dead_code)]
+            impl $ffi_ident {
+                /// Wrap an owned object into a raw pointer.
+                pub(crate) fn new(obj: std::sync::Arc<$T>) -> *const $ffi_ident {
+                    // For unsized types, we need to box the Arc.
+                    Box::into_raw(Box::new($ffi_ident(obj))).cast_const()
+                }
+
+                /// Wrap a borrowed object into a raw pointer.
+                pub(crate) fn new_ref(obj: &std::sync::Arc<$T>) -> *const $ffi_ident {
+                    obj as *const std::sync::Arc<$T> as *const $ffi_ident
+                }
+
+                /// Extract a borrowed reference from a const pointer.
+                pub(crate) fn as_ref<'a>(ptr: *const $ffi_ident) -> &'a std::sync::Arc<$T> {
+                    use vortex::error::VortexExpect;
+                    &unsafe { ptr.as_ref() }
+                        .vortex_expect("null pointer")
+                        .0
+                }
+
+                /// Extract an owned reference from a mutable pointer.
+                pub(crate) fn into_arc(ptr: *const $ffi_ident) -> std::sync::Arc<$T>{
+                    if ptr.is_null() {
+                        vortex::error::vortex_panic!("null pointer");
+                    }
+                    unsafe { Box::from_raw(ptr.cast_mut()) }.0
+                }
+            }
+
+            #[doc = r" Clone a borrowed [`" $ffi_ident "`], returning an owned [`" $ffi_ident "`].\n\n"]
+            #[doc = r" Must be released with [`" $ffi_ident "_free`]."]
+            #[unsafe(no_mangle)]
+            pub unsafe extern "C-unwind" fn [<$ffi_ident _clone>](ptr: *const $ffi_ident) -> *const $ffi_ident {
+                if ptr.is_null() {
+                    vortex::error::vortex_panic!("null pointer");
+                }
+                $ffi_ident::new($ffi_ident::into_arc(ptr.cast_mut()).clone())
```

Comment:
won't this drop the wrapping Box? should this instead be `Arc::clone($ffi_ident::as_ref(ptr))` ?

Reply:
Yes it will, but we don't have another choice in the arc_dyn case. If we just incref the internal arc, then two people are going to try and de-allocate the surrounding box.

In the arc_wrapper case, yes we can just incref: https://github.com/vortex-data/vortex/pull/3425/files/0355896381e0264eb379843e3bf5b0665a86d2e0#diff-5965f190b698e12b6c74e058b0e756b46f95b8adb8c80f6b6c0f91131b4f090eR165

Reply:
It should just make a new box?

```
vx_array *array1; // array1 under the hood is a Box<Arc<dyn Array>>
vx_array* array2 = vx_array_clone(array1);
// now, array1's Box was dropped, which means attempts to access it will be use-after-free
```

Am I reading this wrong?

Reply:
Nope, you're completely right.

Man cast_mut is dangerous. But also, very glad this is centralized now. Won't merge until I add some tests for all these.